Fix TCP transport presence events to work with a NAT

skizunov opened the pull request:
Currently, the presence events do not work when a salt-master is
communicating with salt-minions that are behind a NAT. This is a common
cloud administration use case. It doesn't work for the following reasons:
- The IP address that the salt-master sees is the NAT IP address and it
  is different than the IP address that the salt-minion thinks it has.
  Hence matching by IP address fails.
- The salt-master may see multiple minions through a NAT with the same
  IP address, but different port numbers.

The following challenges have been observed:
- The PUB server connections are the ones that are persistent.
  Unfortunately, the server has no idea which PUB connection corresponds
  to which salt-minion.
- The salt-master can figure out which salt-minion it is communicating
  with it over the REQ server connections. However, these are not
  persistent and there is no way to correlate the REQ Server connections
  with the PUB server connections. The reason is that in the NAT case,
  multiple salt-minions can have the same IP address. And the REQ and PUB
  port numbers will obviously be different.

Due to these challenges, the following solution has been implemented:
- When the minion connects or reconnects the PUB connection to the
  master, it will send data that identifies it to the master over the PUB
  connection. This data is encrypted with AES and includes an identifying
  token, similar to how REQ connections are protected.
- Once the master receives this data and verifies its authenticity, it
  will proceed to fire the correct presence events. Also, when the
  connection is disconnected, it will also proceed to fire the correct
  presence events.
- These presence events will fire from the PUB Server only when the TCP
  transport is the only transport used, and when `presence_events` is
  True in the opts. In this case, it will disable the `presence_events`
  functionality in the `Maintenance` process.

Additional comments:
- Now we also selectively publish to minions (when using tgt_type
  'list') instead of always publishing to all connected minions, since
  we now associate a minion id for each PUB connection. For large numbers
  of minions, publishing to all minions when only a small subset are
  needed greatly increase network traffic and degrade performance.

Signed-off-by: Sergey Kizunov sergey.kizunov@ni.com

skizunov committed:
Fix TCP transport presence events to work with a NAT

Currently, the presence events do not work when a salt-master is
communicating with salt-minions that are behind a NAT. This is a common
cloud administration use case. It doesn't work for the following reasons:

- The IP address that the salt-master sees is the NAT IP address and it
is different than the IP address that the salt-minion thinks it has.
Hence matching by IP address fails.
- The salt-master may see multiple minions through a NAT with the same
IP address, but different port numbers.

The following challenges have been observed:
- The PUB server connections are the ones that are persistent.
Unfortunately, the server has no idea which PUB connection corresponds
to which salt-minion.
- The salt-master can figure out which salt-minion it is communicating
with it over the REQ server connections. However, these are not
persistent and there is no way to correlate the REQ Server connections
with the PUB server connections. The reason is that in the NAT case,
multiple salt-minions can have the same IP address. And the REQ and PUB
port numbers will obviously be different.

Due to these challenges, the following solution has been implemented:
- When the minion connects or reconnects the PUB connection to the
master, it will send data that identifies it to the master over the PUB
connection. This data is encrypted with AES and includes an identifying
token, similar to how REQ connections are protected.
- Once the master receives this data and verifies its authenticity, it
will proceed to fire the correct presence events. Also, when the
connection is disconnected, it will also proceed to fire the correct
presence events.
- These presence events will fire from the PUB Server only when the TCP
transport is the only transport used, and when `presence_events` is
True in the opts. In this case, it will disable the `presence_events`
functionality in the `Maintenance` process.

Additional comments:
- Now we also selectively publish to minions (when using tgt_type
'list') instead of always publishing to all connected minions, since
we now associate a minion id for each PUB connection. For large numbers
of minions, publishing to all minions when only a small subset are
needed greatly increase network traffic and degrade performance.

Signed-off-by: Sergey Kizunov <sergey.kizunov@ni.com>

basepi commented:
This is awesome, @skizunov! My initial read-through reaction is positive, but I'm want to go through this more carefully, since this is pretty core code.

I would also like to get @jacksontj, @thatch45, @cachedout, and @DmitryKuzmenko involved in this review.

joshuaprewitt commented:
This enhancement is extremely important to our companies ability to manage a large distributed collection of systems from a central cloud-based (Amazon or Azure) server.  In addition, I would expect it to be very important for other [IOT](https://en.wikipedia.org/wiki/Internet_of_Things) vendors as well.

joshuaprewitt commented:
Any updates on this pull request?

@jacksontj, @thatch45, @cachedout, and @DmitryKuzmenko

cachedout commented:
@joshuaprewitt We are a little short-handed on pull-request reviewers this week. This should easily make the cutoff for the Boron release though. We'll get it in as soon as we can. Apologies for the delay.e

jacksontj commented on salt/transport/tcp.py:
```diff
@@ -773,38 +812,200 @@ def handle_future(future):
+            return
+
+        clients = self.present[id_]
+        if client not in clients:
```
Might want a trace/debug error here, as this situation should never happen.

skizunov replied:
This situation could happen, since _remove_client_present() can be called from both _stream_read() and publish_payload(). If it is called twice, you will run into this condition. I will go ahead and add a comment explaining this.

jacksontj commented on salt/transport/tcp.py:
```diff
@@ -773,38 +812,200 @@ def handle_future(future):
+                client.close()
+                self._remove_client_present(client)
+                self.clients.discard(client)
+                return
```
I'm fairly certain returns aren't allowed in coroutine methods-- as they are generators. To do this you'd need to do the `raise tornado.gen.Return(None)` or something similar.

skizunov replied:
I wanted to note that "return" is valid here.

http://tornadokevinlee.readthedocs.org/en/latest/gen.html

_By analogy with the return statement, the value argument is optional, but it is never necessary to raise gen.Return(). The return statement can be used with no arguments instead._

I will go ahead and change this to a "break" instead of "return" anyway.

jacksontj replied:
It is in python 3, but not for python 2 (http://tornadokevinlee.readthedocs.org/en/latest/gen.html#tornado.gen.coroutine)

```
. In Python 3.3+, it is also possible for the function to simply use the return value statement (prior to Python 3.3 generators were not allowed to also return values)
```

skizunov replied:
@jacksontj : I believe that even in Python 2.x, `return` by itself should work and `raise gen.Return()` is not necessary. In Python 2.x, `return` followed by a value will not work, and in this case, `raise gen.Return(some_value)` is required. This is no longer required in Python 3.x.

My quote from my previous post above mentions that, and it doesn't qualify Python 2.x vs Python 3.x, so I believe the quote is referring to both. Also, here is another statement supporting my assertion:

http://stackoverflow.com/questions/27020914/return-from-tornado-gen-engine-function-without-yield


_"return" with no argument is the same as "raise gen.Return()". There may be some confusion about Python 2 and 3: in Python 2 a generator can't return a value, but a bare "return" works the same in Python 2 and 3. Can you say more about what you mean, "return still doesn't work"?_

On the other hand, it doesn't hurt anything to have `raise tornado.gen.Return()`, and it is a valid style choice over `return` if consistency is a goal. I am just saying that it isn't absolutely required.

jacksontj commented on salt/transport/tcp.py:
```diff
@@ -773,38 +812,200 @@ def handle_future(future):
     def handle_stream(self, stream, address):
         log.trace('Subscriber at {0} connected'.format(address))
-        self.clients.append((stream, address))
+        client = PubServerClient(stream, address)
```
We might want to rename `PubServerClient`, only because that class makes me think its the client to the pub server-- to be used on a minion. When in reality it for encapsulating a client of the pubserver. Maybe we call it `Subscriber`?

jacksontj commented on salt/transport/tcp.py:
```diff
@@ -773,38 +812,200 @@ def handle_future(future):
+            topic_lst = package['topic_lst']
+            for topic in topic_lst:
+                if topic in self.present:
+                    # Create a separate copy (as a list this time) to iterate
```
Since the publisher runs in a single process, and this next block of code has no `yield` statements, you shouldn't have to make the copy since you are the only coroutine running until you `yield`.

jacksontj commented:
In general, this is awesome-- pretty much as I was planning on doing it, so I can knock it off my TODO list! I put a few comments in there, but in addition to that I really want to see some tests around this new functionality-- especially since its core functionality.

cachedout commented:
@jacksontj As always, thank you for taking the time to review this. Much appreciated.

@skizunov Just keep me posted on this. Thank you!

skizunov committed:
Incorporated feedback from jackstontj

- Added comment in `_remove_client_present` explaining how not
having the client in the list is possible.
- Removed unnecessary copy operations in `publish_payload`.
- Changed `return` to `break` in `_stream_read`. Note that `return`
is technically valid.
- Renamed `PubServerClient` to `Subscriber`.

skizunov commented:
I added another commit which incorporates the feedback from @jacksontj 

As for the additional tests, is that something that I am expected to do or is it something that you guys will take care of?

cachedout commented:
@skizunov We would prefer tests from the author of the code but I'm willing to merge this if we can figure out a plan for test coverage here. Would you like to give writing tests here a shot? If not, I can do it but I am currently traveling on business and it might take me a little while before I can sit down and do it.

skizunov commented:
@cachedout : I am not very familiar with your testing architecture and the types of tests you currently have, so I would prefer if we can figure out a plan for test coverage and merge what we have.

If you already have something that tests `presence_events`, then I would imagine you can just reuse that test and set the transport to `TCP`.

cachedout commented:
@skizunov That's fine. I am willing to merge this without tests for the time being and direct our QA team to start developing a test plan to cover this functionality. Thank you very much for these fixes!

cc: @jfindlay and @ssgward

DmitryKuzmenko commented:
This looks great for me! Thanks @skizunov
